### PR TITLE
Add minify config

### LINKED ISSUE #6621: Keep single/double quotes in `--minify`
`--minify` strips out all superfluous quotes. This breaks Netlify Forms handling, which is checking for `name="field-name"`, and fails on `name=field-name`.

Yes, Netlify should be able to process valid HTML. But, it would be nice to have a setting in Hugo to keep single/double quotes, OR change the defaults to have 'em.

In my opion, minifying HTML should be less aggressive, and be just about removing unnecessary white spaces  and comments.

See also: https://github.com/tdewolff/minify#html

### LINKED ISSUE #6750: Add option to disable inline SVG minification
The `minify` tool used by hugo to minify the output works great on HTML, but there have been several issues with it destroying inlined SVGs:

- https://github.com/tdewolff/minify/issues/275
- https://github.com/tdewolff/minify/issues/284
- https://github.com/tdewolff/minify/issues/269
- https://github.com/gohugoio/hugo/issues/6475

These issues stem from the fact that `minify` changes the svg `<path>` attributes and attempts to optimize the drawing of lines on the svg canvas. This is a very hard problem, and as a result hugo users are affected by SVG optimization bugs from time to time. Furthermore, the testing for SVG minification in the `minify` repo seems to still be rudimentary: https://github.com/tdewolff/minify/blob/master/svg/pathdata_test.go 

In `minify` project readme https://github.com/tdewolff/minify it says:
**"Status: SVG: partially implemented; in development"**

I think in light of this, we should add an option to disable inline SVG minification, or - even better - be able to pass options through to minify if another class of problems emerges.

### PR DESCRIPTION
**Pull request by satotake**
related https://github.com/gohugoio/hugo/issues/6750

This PR enables users to turn on/off by minifier types and to configure `minify` with its options 
This is proposal and any comments and suggestions are welcome

## content
- [mod] ~Update github.com/tdewolff/minify~
- [minifiers] New config.go
- [docshelper] Add func merge DocProviders
- [docs] update docs.json

## related issues

- Fixes #6750 
- partially #6892 (preserve space)
- ~Fixes~ Upstream #6621 

edit:
By downgrading `minify` version, #6621 became an upstream issue

### CODE REVIEW

**Comment by bep** [comment]
This looks really good. I have 3 comments:

1. We cannot update tdewolff deps before we have verified the reason I had to roll back recently (it broke sites badly in the wild, including the Hugo site)

2. I have learned some lessons from Chroma/Blackfriday/Pygments ... there may be a new super minifier on the scene in a year ... with totally different options. So I suggest we split out the general and put the specifics below `tdewolff` (same as in the markup config)

3. I'm not sure about this one, but if I disable XML and do `{ { $xml : $xml | minify }}` what happens?  I assume I get `myxml.min.xml` but with unchanged content. I think we need to split the minifiers in two (one for the final content and the other for the `minify` func) ...?

**Comment by satotake** [comment]
Thanks for comment, @bep 

1 & 2 are done

Please note that #6621 is not solved even by merging this because `minify` is not updated and does not have `KeepQuotes` option.

I admit that I am also not so confident about 3 but let me explain.

>  I disable XML and do { { $xml : $xml | minify }} what happens?

Minification of `resource` raises error and cause build fail when you unset the minfier.
On the other hand, minification of inline contents in `.html` file never raises error.

That is my intention but it is so vague as specification that I have just added a test case and improved the error message.

Second,
> I think we need to split the minifiers in two (one for the final content and the other for the minify func) ...?

It is possible to separate `content` minifier from `resource` minifier in terms of both configurations and processing behaviour (I mean #6724 ).
However I believe that it will be better to use at least single configuration because

1. It reduces configuration simplicity (Hugo's advantage)
2. Most of users who configure minification demand lower level of minification and higher validity of contents. It will be better to apply more secure settings for them in sites' scope.
3. Rquired settings of minification vary with users' situation but they are independent of whether minification targets are derived from `content (or layout)` or `resource`

File: minifiers/config.go [hunk 1]
```diff
@@ -27,54 +27,64 @@ import (
...
-	EnableHtml: true,
-	EnableCss:  true,
-	EnableJs:   true,
-	EnableJson: true,
-	EnableSvg:  true,
-	EnableXml:  true,
-
-	Html: html.Minifier{
+var defaultTdewolffConfig = tdewolffConfig{
+	EnableHTML: true,
```
Comment (bep) [inline minifiers/config.go]: I think we can assume that any future minifier will support all of these fomats, so I suggest we keep all the Enable* on top level.

Comment (satotake) [reply minifiers/config.go]: done!
thanks you for reviewing

**Comment by bep** [comment]
Thanks for this. I think we need to settle 3 before we can wrap this.

```bash
{ { $xml := $xml | minify }} 
```

I _think_ I would expect the above to just silently drop the _minify_ statement if I disable minify, that statement could well live in a theme I don't control.

But you are probably right (?) about having 1 disable for both of these. We can consider adding some flag later if needed.

/cc @kaushalmodi @regisphilibert

**Comment by regisphilibert** [comment]
I can see enabling/disabling is made by Type so it seems very controllable. I'm not familiar with mifnication technology.

I'm simply curious of what happens if a CSS file's minified `.Content` is printeed in an HTML file. I suppose if CSS minification is disabled, but HTML is, then the HTML minification will take care of minifying the CSS?

**Comment by bep** [comment]
> then the HTML minification will take care of minifying the CSS?

Not if you include it in style tags (which I assume you should/would).

**Comment by satotake** [comment]
> I think I would expect the above to just silently drop the minify statement if I disable minify, that statement could well live in a theme I don't control.

You are right @bep . I failed to be careful for `theme`. I will fix

**Comment by bep** [comment]
This looks really good now, thanks a lot. I will pull this and apply some adjustments.

**Comment by bep** [comment]
I will complete this in #7077

**Pull request closed without merging**